core/common/src/main/java/alluxio/collections/IndexedSet.java

```diff
@@ -176,26 +248,40 @@ public boolean add(T object) {
         return false;
       }
 
-      // Update the indexes.
-      for (Map.Entry<FieldIndex<T>, ConcurrentHashMap<Object, ConcurrentHashSet<T>>> fieldInfo :
-          mIndexMap.entrySet()) {
-        // For this field, retrieve the value to index
-        Object fieldValue = fieldInfo.getKey().getFieldValue(object);
-        // Get the index for this field
-        ConcurrentHashMap<Object, ConcurrentHashSet<T>> index = fieldInfo.getValue();
-        ConcurrentHashSet<T> objSet = index.get(fieldValue);
-        if (objSet == null) {
-          index.putIfAbsent(fieldValue, new ConcurrentHashSet<T>());
-          objSet = index.get(fieldValue);
+      if (mIndexMapUnique != null) {
+        for (Map.Entry<FieldIndex<T>, ConcurrentHashMap<Object, T>> fieldInfo : mIndexMapUnique
+            .entrySet()) {
+          // For this field, retrieve the value to index
+          Object fieldValue = fieldInfo.getKey().getFieldValue(object);
+          // Get the index for this field
+          ConcurrentHashMap<Object, T> index = fieldInfo.getValue();
+          // Update the indexes.
+          index.putIfAbsent(fieldValue, object);
```

Comment:
This would happen if multiple objects had the same value for the index, which technically should be illegal, since this is a unique index. Maybe we should throw `IllegalStateException`?